File: src/main/java/io/github/liquidTruffle/runtime/LiquidRuntime.java
package io.github.liquidTruffle.runtime;

import io.github.liquidTruffle.filters.Filter;
import com.oracle.truffle.api.frame.VirtualFrame;
import java.util.Map;
import java.util.List;
import java.util.Collections;

public class LiquidRuntime {
    
    public static boolean isTruthy(Object v) {
        if (v == null) {
            return false;
        } else if (v instanceof Boolean) {
            return (Boolean) v;
        } else if (v instanceof Number) {
            return ((Number) v).doubleValue() != 0.0;
        } else {
            return !v.toString().isEmpty();
        }
    }

    @SuppressWarnings("unchecked")
    public static Map<String, Object> getVariables(VirtualFrame frame) {
        Object[] args = frame.getArguments();
        Object o = args.length > 0 ? args[0] : null;
        if (o instanceof Map) {
            return (Map<String, Object>) o;
        } else {
            return Collections.emptyMap();
        }
    }

    public static Object getVariable(VirtualFrame frame, String name) {
        return getVariables(frame).get(name);
    }

    @SuppressWarnings("unchecked")
    public static Object applyFilter(VirtualFrame frame, String filterName, Object input, List<Object> args) {
        Object[] frameArgs = frame.getArguments();
        Object fObj = frameArgs.length > 1 ? frameArgs[1] : null;
        Map<String, Filter> filters = fObj instanceof Map ? (Map<String, Filter>) fObj : Collections.emptyMap();
        Filter filter = filters.get(filterName);
        if (filter == null) {
            return input;
        }
        return filter.apply(input, args, Collections.emptyMap());
    }
}

File: src/main/kotlin/io/github/liquidTruffle/runtime/LiquidRuntime.kt
package io.github.liquidTruffle.runtime

import io.github.liquidTruffle.filters.Filter
import com.oracle.truffle.api.frame.VirtualFrame

object LiquidRuntime {
	
	fun isTruthy(v: Any?): Boolean {
		return when (v) {
			null -> false
			is Boolean -> v
			is Number -> v.toDouble() != 0.0
			else -> v.toString().isNotEmpty()
		}
	}

	@Suppress("UNCHECKED_CAST")
	fun getVariables(frame: VirtualFrame): Map<String, Any?> {
		val args = frame.arguments
		val o = if (args.isNotEmpty()) args[0] else null
		return if (o is Map<*, *>) o as Map<String, Any?> else emptyMap()
	}

	fun getVariable(frame: VirtualFrame, name: String): Any? {
		return getVariables(frame)[name]
	}

	@Suppress("UNCHECKED_CAST")
	fun applyFilter(frame: VirtualFrame, filterName: String, input: Any?, args: List<Any?>): Any? {
		val frameArgs = frame.arguments
		val fObj = if (frameArgs.size > 1) frameArgs[1] else null
		val filters = if (fObj is Map<*, *>) fObj as Map<String, Filter> else emptyMap()
		val filter = filters[filterName] ?: return input
		return filter.apply(input, args, emptyMap())
	}
}
